### PR TITLE
osd: fix fiemap inconsistency during scrub

### PR DESCRIPTION
**Pull request by mslovy**
Currently, because of some reasons, Although the content in object is consistent,
    the allocation blocks on filesystem may be inconsistent,
    which lead to different filesystem space usage for a same object.
    
    Different space usage for same objects may cause huge inbalance based on current crush alg.
    Therefore, we have enough confidence to consider this case as an inconsistency and report
    a scrub error and repair it.

### CODE REVIEW

**Comment by liewegas** [comment]
Hmm, this makes me nervous.  Different ObjectStore implementations (e.g., FileStore vs BlueStore) might be returning different fiemap values, and in general these are treated as hints and are not well-defined.  I suspect we'll see these "inconsistencies" come up frequently.

What cases are you seeing or concerned about that lead to a material inconsistency?

**Comment by mslovy** [comment]
@liewegas ， for those cluster using default filestore_fiemap = false option.   after rebalance or recovery，it will lead to inconsistency.  Those kind of inconsistency always cause imbalance space usage for a same pg.

I think the logical mapping in bluestore also need to be same for replica objects, even though the physical mapping will be totally different?

**Comment by mslovy** [comment]
Oh, I see.  bluestore  use interval_set<uint64_t>  while filestore and kstore uses map<uint64_t, uint64_t> ... is that possible to unify it?

**Comment by liewegas** [comment]
It's not just that.  If we want a pure canonical definition of the result 
then any 0's or 0'd regions of the object should be "holes," but in 
practice we don't distinguish between writing 0's and not writing to a 
file region. ANd even if we did it is dependent on block sizes, alignment, 
and runtime behavior of the implementation (does it translate 0's into 
holes?  was a hole read and rewritten as 0's?).  It's not well-defined 
what the result should be--it's just a hint.  The most you could probably 
say is that it's an upper bound on what is known to be allocated, but that 
bound may vary across OSDs depending on what objectstore they're using, 
which version they're running, etc.

**Comment by mslovy** [comment]
So do you think it is more suitable to use ceph-object-tool to solve it by user itself?  ceph-object-tool can _zeros the object where it is really 0's, which will be treated differently by different backend (esp. for FileStore, it will be punch zeros on supported filesystem automatically). But it is still inconvenient since ceph-object-tool is a offline tools and it is hard to operation for the clusters > 500 osds and even more.

Another issue is that, since fiemap for filestore is still default false so that it is quite often to make the same object allocating different spaces on different osd, it is a critical problem and I think we should prevent this:
1) for those write to disable fiemap backend,  always allocate the whole object space at once? we can achieve this goals to setting filestore_max_alloc_hint_size = object_size. But filestore_xfs_extsize is default to false because of performance issue and space issue.
2) treat write zeros as holes for those supported systems,  but it is almost impossible to do it since page_size for different system, device varies and many many other stucks.

**Comment by mslovy** [comment]
can we add option such as osd_scrub_detect_fiemap_inconsistency (and default to false)?  So use can decide whether treating it as a problem. (Generally, it is always using the same backend, same hardware in the production environment. Under this situation, it ought to be consistent fiemap among the same objects ).

**Comment by liewegas** [comment]
I'm more concerned with the root cause of the disparity.  I would expect 
that differences in fiemap will be noise (block alignment, essentially a 
rounding error).  If you are seeing that the actual allocation in XFS is 
varying significantly such that real utilization is noticably skewed, we 
should figure out what path is responsible for causing that and fix it.  

...and I just realized that this is happeneing everywhere because fiemap = 
false.  Okay.  The problem is that we have to be extremely careful 
enabling fiemap because it is buggy in so many cases...

My preference is probably to leave it off for filestore so that we 
effectively only get it with BlueStore, but if you have an environment 
where this disparity is problematic and you are confident your kernel is 
correctly handling fiemap, then enable it.  That doesn't really help users 
now, though.  :(

...

Hmm.  To correct the existing imbalance in an existing cluster is tricky.  
Perhaps scrub could have a fiemap threshold such that if the allocation is 
higher on a replica by some factor (>20%, maybe) then it scrub will fix it 
up.  I think if we do anything like this, though, it should be off by 
default, and enabled only on clusters where the admin is trying to clean 
this up.  It should probably also only trigger a repair *after* a 
deep-scrub happens and we verify the objects are in fact all in sync.

**Comment by mslovy** [comment]
Agree, so we can add bool such like allocation_fix in MOSDScrub, which is passed from ceph osd deep-scrub-fix-allocation from command line.
And also use options scrub_allocation_fix_threshold to tell that when we should fix those difference
right?

**Comment by liewegas** [comment]
Why do we need a flag in MOSDScrub?  It seems like this would just be an option controlling the primary's behavior during deep scrub..

**Comment by mslovy** [comment]
Since if I follow the opinion that it is a temporally fix and someone may use it to fix allocation only if they find it is really needed for their systems. So if it is an option, Operator may change option first, and tell osd running deep-scrub, then change option to close the feature again for safety.

So it is better to treat it as a feature, right? Not a temporally fix for an environment and if we enable it, auto-fix will be executed during every deep-scrub?  Then I agree.  And of course, it is much simple to just add an option, which is my preference.

**Comment by mslovy** [comment]
@liewegas ,  https://github.com/ceph/ceph/blob/master/src/osd/PrimaryLogPG.cc#L4377 
here use map<uint64_t, uint64_t> m; 
and also mapext also use map<uint64_t, uint64_t>.
But in BlueStore, we use interval_set<uint64_t>.
This would lead to a decode error I think?
Should we add constructor such as interval_set(const map<T,T> &m) in interval_set.h  or change interval_set in BlueStore fiemap to map<>?

**Comment by mslovy** [comment]
@liewegas ,  https://github.com/ceph/ceph/blob/master/src/osd/PrimaryLogPG.cc#L4377 
here use map<uint64_t, uint64_t> m; 
and also mapext also use map<uint64_t, uint64_t>.
But in BlueStore, we use interval_set<uint64_t>.
This would lead to a decode error I think?
Should we add constructor such as interval_set(const map<T,T> &m) in interval_set.h  or change interval_set in BlueStore fiemap to map<>?

**Comment by mslovy** [comment]
Also add two option to make it default false so that it will not be got a scrub error

**Comment by liewegas** [comment]
map<T,T> and interval_set<T> encode the same.  (a u32 count, following by that many pairs of T values)

**Comment by mslovy** [comment]
Oh, I see. Revert the newest commit

**Comment by mslovy** [comment]
>>> Hmm, this makes me nervous. Different ObjectStore implementations (e.g., FileStore vs BlueStore) >> >>> might be returning different fiemap values, and in general these are treated as hints and are not well->>> defined. I suspect we'll see these "inconsistencies" come up frequently.

>>> What cases are you seeing or concerned about that lead to a material inconsistency?

I think you are right here. Even If the same backend, it may have the different behavior and allocation strategy (we can set different bluestore_min_alloc_size on different osds. Even if it would be stupid under most cases, it is possible.)
One case would be:
ssd primary,  use hdd store replicas so that it leads to different allocation strategy during

So it is also possible for filestore with different filestore_max_alloc_hint_size.

Although this is the concern, suppose it all params are the same for three racks case and failure domain is rack. Then three replica will definitely falls in the three different racks and lead to imbalance among the three racks. In our environment, it appears to highest rack used 85%, lowest rack used 78%,  different with 7% and it would a large difference.

So even though, this is not quite perfect solution,  at least, if we enable those options carefully, it can solve part of those issues. I cannot find better ways to solve this online.

File: src/common/config_opts.h [hunk 1]
```diff
@@ -823,6 +823,8 @@ OPTION(osd_deep_scrub_interval, OPT_FLOAT, 60*60*24*7) // once a week
 OPTION(osd_deep_scrub_randomize_ratio, OPT_FLOAT, 0.15) // scrubs will randomly become deep scrubs at this rate (0.15 -> 15% of scrubs are deep)
 OPTION(osd_deep_scrub_stride, OPT_INT, 524288)
 OPTION(osd_deep_scrub_update_digest_min_age, OPT_INT, 2*60*60)   // objects must be this old (seconds) before we update the whole-object digest on scrub
+OPTION(osd_deep_scrub_allocation_fix, OPT_BOOL, false) // fix object allocation diff among the replicas
+OPTION(osd_deep_scrub_allocation_fix_threshold, OPT_FLOAT, 0.20) // the imbalance thereshold for us to fix the allocation
```
Comment (amitkumar50) [inline src/common/config_opts.h]: @mslovy Wrong spelling "thereshold"

**Pull request closed without merging**